libsolidity/interface/OptimiserSettings.h

```diff
@@ -45,6 +45,7 @@ struct OptimiserSettings
 			"xarulrul"                 // Prune a bit more in SSA
 			"xarrcL"                   // Turn into SSA again and simplify
 			"gvif"                     // Run full inliner
+			"xaTPrci"                  // Prune unused return values and inline
```

Comment:
@cameel Can you help to determine a good sequence for this, perhaps with the help of yul-phaser? I was expecting the gas cost test `test/libsolidity/gasTests/abiv2_optimised.sol` to go down. Maybe there is a better sequence.

Reply:
Well, this sounds like a good opportunity to try out the phaser. Which other input files would you recommend besides `test/libsolidity/gasTests/abiv2_optimised.sol`?

A metric will be a problem though. Right now I only have a code size metric and even with that weights need some tuning. So I'm not sure if this will be good for finding sequences that specifically remove unused return values.

Reply:
Maybe `test/libsolidity/gasTests/abiv2_optimised.sol` is enough to start with.

The code size metric should work, I think.

Reply:
### tl;dr
I tried the phaser today and here's what I found:
- The new optimizer step crashes which interferes with the search. The best sequences I found do not include `P` at all.
-  It's hard to tell if these sequences are actually any better. I think they're just stealing steps that would be used later anyway.
- Phaser is not good at looking for better fragments in the middle of the sequence. I could try to improve it. Adding a `--suffix` option to be able to test with a full sequence might help.


### Running the phaser
```bash
solc --ir test/libsolidity/gasTests/abiv2_optimised.sol | grep -v 'IR:' > abiv2_optimised.yul

yul-phaser abiv2_optimised.yul \
    --show-initial-population \
    --min-chromosome-length 1 \
    --max-chromosome-length 10 \
    --prefix NdhfoDgvulfnTUtnIfxarrscLMcCTUtTOntnfDIulLculVculjjTpeulxarulrulxarrcLgvif
    --random-population 10 \
    --population xaTPrci \
    --population P
```

This creates the initial population out of 10 random optimization sequences plus `xaTPrci`. Also tells the phaser to pre-optimize each program with `NdhfoDgvulfnTUtnIfxarrscLMcCTUtTOntnfDIulLculVculjjTpeulxarulrulxarrcLgvif` and only then apply and rate sequences from the population.

Note:
- `ReasoningBasedSimplifier` is **very** slow. I had to locally remove it from step list in `Chromosome::allStepNames()` to get reasonable speed. With it, every round takes minutes (even on a `RelWithDebInfo` build). Without it, it's less than a second per round.
- Command-line input validations for some options are incomplete so it can sometimes crash on invalid input. For example make sure you always specify the population and do not use step abbreviations that do not exist. I need to make them more robust.

### Phaser output
I did a few runs and selected those that managed to run the longest. Unfortunately there seems to be a bug in this new optimization step and phaser crashes sooner or later (more about it below).

Note that the fitness values are relative to the size of the original program (1000 means no change, 500 means half the size, etc.).

#### First run (without `P` in population)
```
---------- INITIAL POPULATION ----------
353 TcMrsppPtu
541 lT
841 ojTDV
1000 f
1000 vre
1000 hihcdOMgL
1001 eNvihpcnDg
1135 tmPgthU
1166 xaTPrci
1167 IdOaLI
1286 rhUgfarDm
```

```
---------- ROUND 219 [round: 0.2 s, total: 38.2 s] ----------
227 vcrsujOmLTu
227 vcrsujOmTLu
227 vcrsujOmTOu
227 vcrsujOmTTu
227 vcrsujOmThu
227 vcrsujOmTsu
231 vcrsujOmLhu
251 vcrsupOmThu
261 vcrsvjOosTu
540 nslnxsP
663 vhrsuVOmTOx
1000 L
Segmentation fault (core dumped)
```

#### Second run
```
---------- INITIAL POPULATION ----------
290 PLppOmhuON
496 gurcdhtmf
541 lof
1000 P
1000 T
1000 Vo
1013 jToiNIeta
1030 fC
1136 DoNogmtp
1149 OTUim
1166 xaTPrci
1197 acnC
```
```
---------- ROUND 928 [round: 0.3 s, total: 261.1 s] ----------
226 crTvmsuOj
226 crNTvmsuOj
226 crOTvmsuOj
226 crTDvmsuOj
226 crTIvmsuOj
226 crTLvmsuOj
249 crOvvsuuj
583 crTvmslOj
998 NstM
1000 oD
1000 hMrv
1030 C
Segmentation fault (core dumped)
```

#### Optimized programs
If you want to see the optimized programs, you can use this command:
```bash
yul-phaser abiv2_optimised.yul \
    --mode print-optimised-programs \
    --rounds 0 \
    --show-initial-population \
    --prefix NdhfoDgvulfnTUtnIfxarrscLMcCTUtTOntnfDIulLculVculjjTpeulxarulrulxarrcLgvif \
    --population <your sequence> \
```

Here's the output from some of the sequences for convenience:
- [empty-sequence.yul](https://github.com/ethereum/solidity/files/5933474/empty-sequence.yul.txt)
- [vcrsujOmLTu.yul](https://github.com/ethereum/solidity/files/5933475/vcrsujOmLTu.yul.txt)
- [xaTPrci.yul](https://github.com/ethereum/solidity/files/5933476/xaTPrci.yul.txt)
- [crTvmsuOj.yul.txt](https://github.com/ethereum/solidity/files/5933634/crTvmsuOj.yul.txt)
- [PLppOmhuON.yul.txt](https://github.com/ethereum/solidity/files/5933635/PLppOmhuON.yul.txt)
- [nslnxsP.yul](https://github.com/ethereum/solidity/files/5933496/nslnxsP.yul.txt)

### Analysis
First, your sequence got a score of 1166 which means that it actually makes the optimized program longer (at least for this input file). `P` alone is 1000. This means that it will be easily outcompeted and `P` will quickly disappear from the sequence.

If there is a combination of steps where `P` works better, it might reappear but for that I'd have to tweak algorithm settings to make it converge a bit slower - so that it can try more random things before the single best sequence and its mutations take over the whole population.

I see that the best sequences win by reaching for steps from the `"CTUcarrLsTOtfDncarrIulc"  // SSA plus simplify` part. And they mostly don't contain `P`. These other steps seem to have much bigger influence than `P`.

One extra factor to `P`'s disadvantage is the crash - phaser does not crash when `P` is not in the population so naturally the longest runs are likely not to have `P` in the winning sequence.

There are some interesting sequences in the output that do contain `P`, though they are actually just random mutations that happened to be in the population snapshots I grabbed and would disappear in the next round: 
- `nslnxsP` is much shorter mostly because it removes unused functions. Apart from that there are just a handful of places where calculations were replaced by a constant.
- `PLppOmhuON` has a lot more removed and simplified but it's hard for me to tell how much of it is due to `P`.

Overall, I think that finding a good sequence for this with phaser won't be possible without some modifications:
- A `--suffix` option (similar to the current `--prefix`) would prevent phaser from trying to include steps from the later part of our current sequence.
- Maybe some constraints on which steps should be used.
- Having a better metric might help too.

### Crash
This reproduces it immediately for me:
```
yul-phaser test/libyul/yulOptimizerTests/fullSuite/aztec.yul --population 'P'
```
Here's the relevant part of the stack trace:
```
Program received signal SIGSEGV, Segmentation fault.

(...)

#0  FindFunctionsWithUnusedReturns::operator() (this=0x7fffffffd470, _v=...) at /usr/include/c++/10.2.0/variant:1554

(...)

#17 solidity::yul::ASTWalker::operator() (this=this@entry=0x7fffffffd470, _block=...) at /projects/solidity/libyul/optimiser/ASTWalker.cpp:89
#18 0x000055555568df81 in FindFunctionsWithUnusedReturns::functions (_dialect=..., _references=..., _block=...) at /projects/solidity/libyul/optimiser/UnusedFunctionReturnParameterPruner.cpp:168
#19 0x000055555568e5d8 in solidity::yul::UnusedFunctionReturnParameterPruner::run (_context=..., _ast=...) at /projects/solidity/libyul/optimiser/UnusedFunctionReturnParameterPruner.cpp:187
#20 0x00005555555f7ee2 in solidity::phaser::Program::applyOptimisationSteps (_dialect=..., _nameDispenser=..., _ast=std::unique_ptr<solidity::yul::Block> = {...}, _optimisationSteps=...) at /projects/solidity/tools/yulPhaser/Program.cpp:206
#21 0x00005555555f7fb9 in solidity::phaser::Program::optimise (this=this@entry=0x7fffffffdc80, _optimisationSteps=std::vector of length 1, capacity 1 = {...}) at /usr/include/c++/10.2.0/bits/unique_ptr.h:172

(...)
```

So it basically happens several levels deep in the AST walker code that runs when `FindFunctionsWithUnusedReturns` gets created and its `operator ()` called.

Reply:
That `--suffix` option should be a pretty simple modification so let me know when you deal with the crash. I'll give it another shot.